File: 74.搜索二维矩阵.py
#
# @lc app=leetcode.cn id=74 lang=python3
#
# [74] 搜索二维矩阵
#

# @lc code=start
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        length = len(matrix) * len(matrix[0])
        i, j = 0, length
        while i < j:
            m = (i + j) // 2
            x = m // len(matrix[0])
            y = m % len(matrix[0])
            if matrix[x][y] == target:
                return True
            elif matrix[x][y] < target:
                i = m + 1
            else:
                j = m
        return False
# @lc code=end

File: 74.搜索二维矩阵.cpp
/*
 * @lc app=leetcode.cn id=74 lang=cpp
 *
 * [74] 搜索二维矩阵
 */
#include<vector>
using namespace std;
// @lc code=start
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int i = 0;
        int j = m * n;
        int mid;
        int x, y;
        while(i < j){
            mid = (i + j) >> 1;
            x = mid / n;
            y = mid % n;
            if(matrix[x][y] == target) return true;
            else if(matrix[x][y] < target) i = mid + 1;
            else j = mid;
        }
        return false;
    }
};
// @lc code=end

